### PR TITLE
8265222: revisit foreign library loading

### PR DESCRIPTION
**Pull request by sundararajana**
LibraryLookup removed. CLinker.findNative added. System.loadLibrary has to be used to load library

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

### Issue
 * [JDK-8265222](https://bugs.openjdk.java.net/browse/JDK-8265222): revisit foreign library loading

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.java.net/census#mcimadamore) (@mcimadamore - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/panama-foreign pull/526/head:pull/526` \
`$ git checkout pull/526`

Update a local copy of the PR: \
`$ git checkout pull/526` \
`$ git pull https://git.openjdk.java.net/panama-foreign pull/526/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 526`

View PR using the GUI difftool: \
`$ git pr show -t 526`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/panama-foreign/pull/526.diff">https://git.openjdk.java.net/panama-foreign/pull/526.diff</a>

</details>

### CODE REVIEW

File: src/java.base/windows/native/libjava/jni_util_md.c [hunk 1]
```diff
@@ -36,31 +36,6 @@ void* getProcessHandle() {
     return (void*)GetModuleHandle(NULL);
 }
 
-/*
- * Windows doesn't have an RTLD_DEFAULT equivalent, so in stead we have to
- * iterate over all the modules loaded by the process to implement the
- * default library behaviour.
- */
-void* findEntryInProcess(const char* name) {
```
Comment (JornVernee) [inline src/java.base/windows/native/libjava/jni_util_md.c:44]: I think removing this also allows the removal of `#include <psapi.h>` from the includes. I'll give it a try here as well.

File: test/jdk/java/foreign/libStdLibTest.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+#include <time.h>
+
+EXPORT char *libc_strcat(char *str1, const char *str2);
+EXPORT int libc_strcmp(const char *str1, const char *str2);
+EXPORT size_t libc_strlen(const char *str);
+EXPORT int libc_puts(const char *str);
+EXPORT struct tm *libc_gmtime(const time_t *timer);
+EXPORT void libc_qsort(void *base, size_t nitems, size_t size, int (*compar)(const void *, const void*));
+EXPORT int libc_rand(void);
+EXPORT int vprintf(const char *format, va_list arg);
```
Comment (JornVernee) [inline test/jdk/java/foreign/libStdLibTest.h:43]: This declaration of `vprintf` gives an error on Windows, is it needed? The test passes without as well.  (should came of stdio.h I'd think)

```
test\jdk\java\foreign\libStdLibTest.h(43): error C2375: 'vprintf': redefinition; different linkage
c:\progra~2\wi3cf2~1\10\include\100183~1.0\ucrt\stdio.h(738): note: see declaration of 'vprintf'
   ... (rest of output omitted)
```

File: src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java [hunk 3]
```diff
@@ -126,6 +129,20 @@ static CLinker getInstance() {
         return SharedUtils.getSystemLinker();
     }
 
+    @CallerSensitive
+    public static Optional<MemoryAddress> findNative(String name) {
```
Comment (mcimadamore) [inline src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java:133]: Security manager checks should be removed here. Also, the method is missing javadoc.

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/SharedUtils.java [hunk 4]
```diff
@@ -415,13 +414,12 @@ static MethodHandle wrapWithAllocator(MethodHandle specializedHandle,
...
-        static final LibraryLookup LOOKUP = LibraryLookup.ofDefault();
-
-        static final MethodHandle MH_MALLOC = getSystemLinker().downcallHandle(LOOKUP.lookup("malloc").get(),
+        private static final CLinker linker = getSystemLinker();
+        static final MethodHandle MH_MALLOC = linker.downcallHandle(CLinker.findNative("malloc").get(),
                         MethodType.methodType(MemoryAddress.class, long.class),
                 FunctionDescriptor.of(C_POINTER, C_LONG_LONG));
 
-        static final MethodHandle MH_FREE = getSystemLinker().downcallHandle(LOOKUP.lookup("free").get(),
+        static final MethodHandle MH_FREE = linker.downcallHandle(CLinker.findNative("free").get(),
```
Comment (JornVernee) [inline src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/SharedUtils.java:422]: These symbol lookups fail on Windows.

**Review by mcimadamore — COMMENTED** [review]
Looks like a good simplification! Some comments added.

File: src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java (hunk 3 above)
Comment (sundararajana) [reply src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java:133]: Will remove security manager checks and add javadoc comment

File: src/java.base/windows/native/libjava/jni_util_md.c (hunk 1 above)
Comment (JornVernee) [reply src/java.base/windows/native/libjava/jni_util_md.c:44]: Confirmed, removing the include is safe.

**Comment by mcimadamore** [comment]
Microbenchmarks also need updating - otherwise build fails here:

```
test/micro/org/openjdk/bench/jdk/incubator/foreign/CallOverheadHelper.java:28: error: cannot find symbol
import jdk.incubator.foreign.LibraryLookup;
                            ^
  symbol:   class LibraryLookup
  location: package jdk.incubator.foreign
test/micro/org/openjdk/bench/jdk/incubator/foreign/StrLenTest.java:30: error: cannot find symbol
import jdk.incubator.foreign.LibraryLookup;
                            ^
  symbol:   class LibraryLookup
  location: package jdk.incubator.foreign
test/micro/org/openjdk/bench/jdk/incubator/foreign/Upcalls.java:26: error: cannot find symbol
import jdk.incubator.foreign.LibraryLookup;
                            ^
  symbol:   class LibraryLookup
  location: package jdk.incubator.foreign
   ... (rest of output omitted)
```

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/SharedUtils.java (hunk 4 above)
Comment (JornVernee) [reply src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/SharedUtils.java:422]: Following fix works:

```
diff --git a/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/SharedUtils.java b/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/SharedUtils.java
index d705d414c0d..708c005d1d9 100644
--- a/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/SharedUtils.java
+++ b/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/SharedUtils.java
@@ -50,6 +50,7 @@ import java.lang.invoke.MethodType;
 import java.lang.invoke.VarHandle;
 import java.lang.ref.Reference;
 import java.nio.charset.Charset;
+import java.nio.file.Path;
 import java.util.List;
 import java.util.Map;
 import java.util.function.Consumer;
@@ -415,6 +416,9 @@ public class SharedUtils {
     private static class AllocHolder {

         private static final CLinker linker = getSystemLinker();
+        static {
+            System.load(Path.of(System.getenv("SystemRoot"), "System32", "msvcrt.dll").toAbsolutePath().toString());
+        }
         static final MethodHandle MH_MALLOC = linker.downcallHandle(CLinker.findNative("malloc").get(),
                         MethodType.methodType(MemoryAddress.class, long.class),
                 FunctionDescriptor.of(C_POINTER, C_LONG_LONG));
```

File: src/java.base/windows/native/libjava/jni_util_md.c (hunk 1 above)
Comment (sundararajana) [reply src/java.base/windows/native/libjava/jni_util_md.c:44]: Thanks. Will remove that include.

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/SharedUtils.java (hunk 4 above)
Comment (sundararajana) [reply src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/SharedUtils.java:422]: Yes, will add this as temporary hack for Windows.

File: src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java [hunk 5]
```diff
@@ -129,14 +129,23 @@ static CLinker getInstance() {
         return SharedUtils.getSystemLinker();
     }
 
+
+    /**
+     * Looks up a symbol with the given name in the native libraries loaded by the caller's classloader.
```
Comment (mcimadamore) [inline src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java:134]:
```suggestion
     * Finds the address of a symbol with given name in one of the native libraries associated with the caller's classloader (that is, libraries loaded using {@link System#loadLibrary} or {@link System#load}).
```

File: src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/package-info.java [hunk 6]
```diff
@@ -118,7 +117,7 @@
       }
  * }</pre>
  *
- * Here, we lookup the {@code strlen} symbol in the <em>default</em> library lookup (see {@link jdk.incubator.foreign.LibraryLookup#ofDefault()}).
+ * Here, we lookup the {@code strlen} symbol. C runtime library should have been loaded earlier using {@link System#loadLibrary(String)}.
```
Comment (mcimadamore) [inline src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/package-info.java:120]:
```suggestion
 * Here, we lookup the {@code strlen} symbol; this assumes that a library containing the `strlen` symbol has been loaded earlier, e.g. using {@link System#loadLibrary(String)}.
```

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/SharedUtils.java [hunk 7]
```diff
@@ -415,6 +416,15 @@ static MethodHandle wrapWithAllocator(MethodHandle specializedHandle,
     private static class AllocHolder {
 
         private static final CLinker linker = getSystemLinker();
+        // FIXME: This should go away. This is temporary hack to get testing on Windows going.
```
Comment (mcimadamore) [inline src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/SharedUtils.java:419]: Move the fixme closer to the hack, right before the `if` in the static block

File: src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java [hunk 8]
```diff
@@ -129,14 +129,23 @@ static CLinker getInstance() {
...
+     *
+     * <p>
+     * This method is <a href="package-summary.html#restricted"><em>restricted</em></a>.
+     * Restricted method are unsafe, and, if used incorrectly, their use might crash
+     * the JVM or, worse, silently result in memory corruption. Thus, clients should refrain from depending on
+     * restricted methods, and use safe and supported functionalities, where possible.
+     *
+     * @param name the symbol name.
```
Comment (mcimadamore) [inline src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java:142]:
```suggestion
     * @param the name of the symbol to be searched.
```

File: src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java [hunk 9]
```diff
@@ -129,14 +129,23 @@ static CLinker getInstance() {
...
+     * @return the memory address associated with the library symbol (if any).
```
Comment (mcimadamore) [inline src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java:143]:
```suggestion
     * @return the address of a symbol with given name in one of the native libraries associated with the caller's classloader (if any).
```

**Review by mcimadamore — COMMENTED** [review]
Looks good - added some javadoc suggestions

File: src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java (hunk 8 above)
Comment (sundararajana) [reply src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java:142]: will do

File: src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java (hunk 9 above)
Comment (sundararajana) [reply src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java:143]: will do

File: src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java (hunk 5 above)
Comment (sundararajana) [reply src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java:134]: will change

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/SharedUtils.java (hunk 7 above)
Comment (sundararajana) [reply src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/SharedUtils.java:419]: will do

File: src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/package-info.java (hunk 6 above)
Comment (sundararajana) [reply src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/package-info.java:120]: will fix

File: src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/package-info.java [hunk 10]
```diff
@@ -117,7 +117,9 @@
       }
  * }</pre>
  *
- * Here, we lookup the {@code strlen} symbol. C runtime library should have been loaded earlier using {@link System#loadLibrary(String)}.
+ * Here, we lookup the {@code strlen} symbol; this assumes that a library containing the `strlen` symbol has been loaded
```
Comment (mcimadamore) [inline src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/package-info.java:120]:
```suggestion
 * Here, we lookup the {@code strlen} symbol; this assumes that a library containing the {@code strlen} symbol has been loaded
```

File: src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java [hunk 11]
```diff
@@ -131,16 +131,18 @@ static CLinker getInstance() {
...
      *
      * <p>
      * This method is <a href="package-summary.html#restricted"><em>restricted</em></a>.
      * Restricted method are unsafe, and, if used incorrectly, their use might crash
      * the JVM or, worse, silently result in memory corruption. Thus, clients should refrain from depending on
      * restricted methods, and use safe and supported functionalities, where possible.
      *
-     * @param name the symbol name.
-     * @return the memory address associated with the library symbol (if any).
+     * @param name of the symbol to be searched.
```
Comment (mcimadamore) [inline src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java:143]:
```suggestion
     * @param name the name of the symbol to be searched.
```

**Review by mcimadamore — COMMENTED** [review]
Some more javadoc comments

**Review by mcimadamore — APPROVED** [review]

**Pull request closed without merging**